pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -240,6 +253,26 @@ def add_close_callback(self, callback):
         """
         self._close_callbacks.append(callback)
 
+    def activate_ordering_keys(self, ordering_keys):
+        """Send the next message in the queue for each of the passed-in
+        ordering keys, if they exist. Clean up state for keys that no longer
+        have any queued messages.
+
+        There is enough load capacity to send one message per ordering key
+        because each key is the result of an ack or nack. Since the load went
+        down by one message, it's safe to send the user another message for the
```

Comment:
Similar remark here - what if the load is too high due to the total message size, not count?
This seems to break the assumption in the docstring...

Re: "what happens" - the 11th ordering keys message is normally put into `MessagesOnHold` (in the streaming pull manager's `_on_response()` method). Then they just wait there to be released until the load allows for it.

Reply:
When a message is dropped (see the dispatcher's drop() method) as a result of ack, nack, or drop by the user, messages get released in two ways:
1) dispatcher.drop() calls activate_ordering_keys(), which releases the next message for each ordering key (if they exist). Since at least one message was dropped for each key, it's safe to release a message for each key.
2) dispatcher.drop() then calls maybe_resume_consumer(), which tells the MessagesOnHold class to release the next messages in the queue (within load parameters). These messages may be ordered or unordered.

In your example, the message with the 11th ordering key would be released in step 2.

Reply:
Response to Peter's question:
You're right that it's possible that the next ordered message pulled for the same key is too big, and that load would exceed the max. I should document this in the docstring.

I think this is permissible because the alternative would complicate MessagesOnHold even more. To solve this issue, we'd have to keep track of ordering keys that need to be "activated" next. When MessagesOnHold.get() is called, we'd need to check that side list, and also the main queue of held messages.

The reason why I went with this way of storing ordered messages is to prevent a situation where we have to choose between various ordering keys and the general unordered queue. Fairness would only be guaranteed by random weighting, which is usually linear or, in a more complicated form, log(n) in the number of weights (keys). We'd run into this same issue with this little side list of "ordering keys that need to be activated".

Reply:
Can't say if treating `FlowControl.max_bytes` only as a soft limit is permissible or not, but yes, let's at least document that possibility here.

Sometimes releasing some of the message a bit too early is IMO still a lesser evil than starving.